src/lib/config/ModelConfig.cc

```diff
@@ -0,0 +1,1007 @@
+      ref = configTree_.ref(id).append_child() << ryml::key(key);
+      // Set any appropriate ryml::NodeRef types
+      if (child.is_map()) {
+        ref |= ryml::MAP;
+      }
+      if (child.is_seq()) {
+        ref |= ryml::SEQ;
+      }
+    } else {
+      ref = configTree_.ref(id)[child.key()];
+    }
+    if (child.is_map()) {
+      // If the config option had children, iterate through them.
+      recursiveAdd(child, ref.id());
+    } else if (child.is_seq()) {
+      // If the config option is a sequence, then add the sequence of values
+      // held within the config option (its children) as children to the current
+      // ryml::Tree node identified by `id`
+      ref.clear_children();
+      for (size_t entry = 0; entry < child.num_children(); entry++) {
+        ref.append_child();
+        ref[entry] << child[entry].val();
+      }
+    } else {
+      // If the config option is neither a map nor a sequence, simply add its
+      // value to the ryml::Tree node reference
+      ref << child.val();
+    }
+  }
+}
+
+void ModelConfig::setExpectations(bool isDefault) {
+  // Reset expectations
+  expectations_ = {};
+
+  // Core
+  expectations_.addChild(ExpectationNode::createExpectation("Core"));
+
+  if (isa_ == ISA::AArch64)
+    expectations_["Core"].addChild(
+        ExpectationNode::createExpectation<std::string>("AArch64", "ISA"));
+  else if (isa_ == ISA::RV64)
+    expectations_["Core"].addChild(
+        ExpectationNode::createExpectation<std::string>("rv64", "ISA"));
+  expectations_["Core"]["ISA"].setValueSet(
+      std::vector<std::string>{"AArch64", "rv64"});
+
+  // Early check on [Core][ISA] as its value is needed to inform the
+  // expectations of other config options
+  if (!isDefault) {
+    ValidationResult result = expectations_["Core"]["ISA"].validateConfigNode(
+        configTree_["Core"]["ISA"]);
+    std::string ISA = configTree_["Core"]["ISA"].as<std::string>();
+    if (!result.valid) {
+      std::cerr << "[SimEng:ModelConfig] Invalid ISA value of \"" << ISA
+                << "\" passed in config file due to \"" << result.message
+                << "\" error. Cannot continue with config validation. Exiting."
+                << std::endl;
+      exit(1);
+    }
+    // Set isa_
+    if (ISA == "AArch64") {
+      isa_ = ISA::AArch64;
+    } else if ("rv64") {
+      isa_ = ISA::RV64;
+    }
+  }
+  createGroupMapping();
+
+  expectations_["Core"].addChild(
+      ExpectationNode::createExpectation<std::string>("emulation",
+                                                      "Simulation-Mode"));
+  expectations_["Core"]["Simulation-Mode"].setValueSet(
+      std::vector<std::string>{"emulation", "inorderpipelined", "outoforder"});
+
+  const float clockFreqUpperBound = 10.f;
+  expectations_["Core"].addChild(
+      ExpectationNode::createExpectation<float>(1.f, "Clock-Frequency"));
+  expectations_["Core"]["Clock-Frequency"].setValueBounds(0.f,
```

Comment:
Does a minimum clock frequency of 0GHz make sense? Perhaps the lower-bound should be 1mhz or 0.001 GHz?